include/trieste/wf.h

```diff
@@ -155,13 +161,82 @@ namespace trieste
       {
         return rand();
       }
+      
+      private:
+      Nodes get_symbols(Token type, Node scope)
+      {
+        Nodes symbols;
+        if (scope)
+        {
+          // Look up symbols in the current scope that match the binding type
+          scope->get_symbols(symbols, [&](auto& n) {
+            return (
+              n->type() & flag::lookup &&
+              binding_keys[n->type()].first == type);
+          });
+        }
+        return symbols;
+      }
 
-      Location location(Node n)
+      // Returns true if we should generate a bound variable
+      bool gen_bound(Node parent, Token child_type)
       {
-        return gloc(rand, n);
+        if (!gen_bound_vars)
+          return false;
+        // Check if there is a binding key, e.g. any binding key match the child
+        // type
+        std::vector<Token> bound_nodes = {};
+        for (auto& [parent_type, binds_to] : binding_keys)
+        {
+          if (binds_to.first == child_type)
+            bound_nodes.push_back(parent_type);
+        }
+        if (bound_nodes.empty())
+        {
+          return false;
+        }
+        else if (
+          std::find(bound_nodes.begin(), bound_nodes.end(), parent->type()) !=
+          bound_nodes.end())
+        {
+          // Check if the node we are generating is the binding key of the
+          // current parent node
+          auto key_field_index = binding_keys[parent->type()].second;
+          auto key_field_type = binding_keys[parent->type()].first;
+          auto current_index = parent->size() - 2; // Child is last pushed node
+          if (key_field_type == child_type)
+            // Generate a bound variable if we are not at the binding site
+            return key_field_index != current_index;
+          else
+            return false; 
+        }
+        else // Not at binding site
+          return true;
       }
-    };
 
+      public:
+        Location location(Node parent, Node child)
+        {
+          auto type = child->type();
+          Nodes symbols = get_symbols(type, parent->scope());
+
+          // Ensure we're not always use bindings when they exist by adding +1
+          auto rand_symbol = static_cast<size_t>(next() % (symbols.size() + 1));
+          // If we have a bound symbol to use, we're not at the binding site,
+          // and the parent type allows shadowing, use a bound symbol
+          if (
+            rand_symbol < symbols.size() && gen_bound(parent, type) &&
+            !(parent->type() & flag::shadowing))
```

Comment:
Shouldn't this last thing be checked inside `gen_bound`? The `parent` here is the parent of the node whose location we're generating, but the shadowing check should only come into play when we happen to be generating a bound variable where a new (shadowing) binding should go.